### ISSUE TITLE
create a browser handler for browseURL

### ISSUE THREAD

**Issue by aronatkins** [rstudio/shiny#1850]
When Shiny applications are deployed to a remote server, the `browseURL` function still attempts to use the default URL opener. On some Linux systems, this is `xdg-open`.

The `browseURL` function can be disabled by configuring the `browser` option to `"false"`:
```r
option(browser = "false")
```

Shiny should offer a handler for the `browser` option that broadcasts the URL to the client and then opens that URL in a new tab.

```r
shinyBrowseURLHandler <- function(url) { ... }
option(browser = shinyBrowseURLHandler)
```

Shiny could install this handler automatically (possibly when non-interactive?), but that might raise backwards-compatibility concerns.

**Comment by Dripdrop12** [comment]
Yeah, I'm trying to customize the new bookmarking feature to open in a new tab, and this would be incredibly useful.

**Comment by brockh** [comment]
I agree, browseURL when deployed in a shiny app on shiny server should broadcast to the client.

**Comment by ecarosati** [comment]
I tried a shiny app with Rstudio Connect and I confirm that the call to browseURL does not work, and the message "Cannot visit http://.... because the browseURL function is disabled" appears

**Comment by sound118** [comment]
can someone help with this issue? it's been around for a few years already!!

**Comment by RiboRings** [comment]
Hey! Any updates on this? It would be really helpful for my deployed shiny app too :)

**Comment by gadenbuie** [comment]
Since this issue was opened, Connect has added lint warnings for `browseURL()` and `browse()` at deploy time, so we now actively discourage their use in Shiny apps. Rather than trying to intercept or redirect these calls, the more reliable approach is to move the URL-opening behavior to the client.

A few patterns that work well today:

1. **Custom message + `window.open()`**. Send the URL from the server using `session$sendCustomMessage()`, then open it client-side with `window.open(url, "_blank", "noopener")`. This keeps browser behavior in the browser.

   ```r
   # In your UI
   tags$script(HTML("
     Shiny.addCustomMessageHandler('open-url', function(url) {
       window.open(url, '_blank', 'noopener');
     });
   "))

   # In your server
   observeEvent(input$go, {
     session$sendCustomMessage("open-url", "https://example.com")
   })
   ```

2. **A plain `<a target="_blank">` link:** If the URL is known at render time and opened via a user click, a standard anchor tag with `target="_blank"` is the simplest and most reliable option — no extra messaging needed.

   ```r
   tags$a(
     href = "https://example.com",
     target = "_blank",
     rel = "noopener noreferrer",
     "Open link"
   )
   ```

3. **`shinyjs::runjs()`:** If your app already uses `shinyjs`, you can trigger `window.open()` from an observer with minimal setup.

   ```r
   observeEvent(input$go, {
     shinyjs::runjs("window.open('https://example.com', '_blank', 'noopener');")
   })
   ```

One thing to keep in mind with options 1 and 3: browsers may block `window.open()` calls not directly tied to a user gesture, so anchoring the action to a button or link click is the safest bet.

Given all of this, I'm going to close this issue — but hopefully one of the above gets you where you need to go!
